src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs

```diff
@@ -588,9 +588,46 @@ public void AddLabel(string label, VertexType vertexType, int index)
                 default:
                     return;
             }
-            LabelPlaces.Add(new Tuple<string, Vector3>(label,position));
+            LabelPlaces.Add(new Tuple<string, Vector3>(label, position));
         }
 
+        /// <summary>
+        /// Adds a label to the render package, but first transforms the label by the transform matrix of the 
+        /// relevant graphicItem.
+        /// </summary>
+        /// <param name="label">label</param>
+        /// <param name="vertexType">type of vertex</param>
+        /// <param name="vertIndex">vertex index for base label position</param>
+        /// <param name="instanceIndex">index to use for transform matrix</param>
+        /// <param name="BaseTessellationId">baseTessellation Id of the item this label belongs to.
+        /// Aids in lookup of the correct transform matrix.</param>
+        void IRenderInstancedLabels.AddInstancedLabel(string label, VertexType vertexType, int vertIndex, int instanceIndex, Guid BaseTessellationId)
+        {
+            Vector3 position = new Vector3();
+            switch (vertexType)
+            {
+                case VertexType.MeshInstance:
+                    //
+                    if (vertIndex > mesh.Positions.Count) { return; }
+                    position = Mesh.Positions[vertIndex - 1];
+                    break;
+                case VertexType.LineInstance:
+                    //
+                    if (vertIndex > Lines.Positions.Count) { return; }
+                    position = Lines.Positions[vertIndex - 1];
+                    break;
+            }
+            if (instanceTransforms.ContainsKey(BaseTessellationId) && instanceIndex <= instanceTransforms[BaseTessellationId].Count)
+            {
+                var transform = instanceTransforms[BaseTessellationId]?[instanceIndex - 1];
+                var transformedLabelPos = transform?.ToMatrix3D().Transform(position.ToPoint3D()).ToVector3();
```

Comment:
Any idea how long this takes.  I guess I am curious perf vs within the AddInstance call where you have access to position info directly.  This is also probably very fast.

Reply:
@saintentropy testing with a stopwatch this appears to take 30 - 50 ms for 125000 labels.

Reply:
![Screen Shot 2022-05-23 at 10 49 02 AM](https://user-images.githubusercontent.com/508936/169846908-05372b5d-2969-4036-8dc3-ba4e36d32fad.png)